sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -1720,6 +1736,23 @@ public Object restriction() {
     doFnInvoker.invokeOnTimer(timerId, timerFamilyId, onTimerContext);
   }
 
+  private <K> void processOnWindowExpiration(Timer<K> timer) {
+    try {
+      currentKey = timer.getUserKey();
+      currentTimer = timer;
+      Iterator<BoundedWindow> windowIterator =
```

Comment:
`for (BoundedWindow currentWindow : timer.getWindows()) { ... }` ?

Reply:
currentWindow is defined outside of the method. I believe for-each loops are limited to defining the variable within the `for (...)` so I can't do `for (currentWindow : timer.getWindows()) { ... }`

Reply:
Ah yes, I see that you are mutating a field so that the nested instance can carry it to the invoker. Gotta say there's quite a lot of implicit data flows here. I understand it is mutation for efficiency. Might be worth thinking about how to make them more explicit and local.